subsystems/DrivetrainBase.java

```diff
@@ -75,13 +143,165 @@ public void tankDriveVoltsReverse(double leftVolts, double rightVolts) {
+    }
+
+    // converts rotations per minute (RPM) to meters per second (MPS)
+    private double rpmToMPS(double rotationsPerMinute) {
+        double radiansPerSecond = Units.rotationsPerMinuteToRadiansPerSecond(rotationsPerMinute);
+        double linearMetersPerSecond = radiansPerSecond * (constants.wheelDiameterMeters / 2);
+        return linearMetersPerSecond;
+    }
+
+    public double getLeftVelocityMPS() {
+        return rpmToMPS(getLeftVelocityRPM());
+    }
+
+    public double getRightVelocityMPS() {
+        return rpmToMPS(getRightVelocityRPM());
+    }
+
+    // gets the average velocity of the robot by taking the mean of both sides
+    public double getAbsoluteAverageVelocityMPS() {
+        return (Math.abs(getRightVelocityMPS()) + Math.abs(getLeftVelocityMPS())) / 2;
+    }
+
     // gets the direction of the robot
     public double getHeading() {
         return ahrs.getAngle() * (invertGyro ? -1 : 1);
         // return gyro.getAngle();
     }
 
-    // the next two methods make the class PIDSubsystem compliant. 
+    public double getHeading360() {
+        return Math.IEEEremainder(getHeading(), 360.0/* d */);
+    }
+
+    public double getHeading180() {
+        var heading = getHeading() % 360;
+        if (heading > 180) {
+            return heading - 360;
+        } else if (heading < -180) {
+            return heading + 360;
+        } else {
+            return heading;
+        }
+    }
+
+    public Pose2d getPose() {
+        return odometry.getPoseMeters();
+    }
+
+    public DifferentialDriveWheelSpeeds getWheelSpeeds() {
+        return new DifferentialDriveWheelSpeeds(getLeftVelocityMPS(), getRightVelocityMPS());
+    }
+
+    public DifferentialDriveWheelSpeeds getWheelSpeedsReverse() {
+        return new DifferentialDriveWheelSpeeds(-getLeftVelocityMPS(), -getRightVelocityMPS());
+    }
+
+    public Pose2d getClosestPoseAtTime(double time) {
+        return poseHistory.getClosest(time);
+    }
+
+    // resets the positions of the encoders to 0
+    public void resetEncoders() {
+        leftLeader.setSelectedSensorPosition(0);
+        rightLeader.setSelectedSensorPosition(0);
+    }
+
+    // resets the heading of the robot
+    public void resetHeading() {
+        ahrs.reset();
+    }
+
+    // resets both the heading and the encoders, as well as the odometry of the robot
+    public void resetOdometry() {
+        resetHeading();
+        resetEncoders();
+        odometry.resetPosition(new Pose2d(), new Rotation2d());
+    }
+
+    // resets the odometry to a given position param
+    public void resetOdometry(Pose2d pose) {
+        resetEncoders();
+        odometry.resetPosition(pose, ahrs.getRotation2d());
+    }
+
+    public double getAverageEncoderDistance() {
+        return (leftLeader.getSelectedSensorPosition() + rightLeader.getSelectedSensorPosition()) / 2;
+    }
+
+    public double getTurnRate() {
+        return -ahrs.getRate();
```

Comment:
unsure if this has to be negative - maybe at a todo comment to check this

Reply:
it may be the difference between some things having a default clockwise vs counter-clockwise rotation. But this should be made positive, and def leave a TODO there so we can check it later